### PR TITLE
Use correct heap image when setting docker limits

### PR DESCRIPTION
**Pull request by hoshsadiq**
Due to the current way JVM is implemented, Java's default `MaxHeapSize` does not get calculated properly, in that, it does not read the limits set by cgroups, and as such, does not listen to docker when setting a maximum memory size or cpu limit. This solution has been solved over and over again, and this just reuses current solutions (specifically from [fabric8's Java image](https://github.com/fabric8io-images/java)). The additional scripts simply calculate an appropriate `MaxHeapSize` and this then gets added into `JAVA_OPTS`. This allows for improved docker handling of Jenkins and in fact runs it more stable. Consequently, Kubernetes will be able to handle it correctly as well.

Reproducing this issue is simple, run below docker commands on current Jenkins image
```
docker run --name jenkins  -e JAVA_OPTS='-XX:+PrintFlagsFinal -XX:+PrintGCDetails' -m=250m -d jenkins
docker logs jenkins 2>/dev/null | grep -i MaxHeapSize
```
You will see that the `MaxHeapSize` calculated is way beyond what the memory limit is. As such, this causes an OOM should Jenkins start using more memory (see below).
```
    uintx MaxHeapSize                              := 524288000                           {product}
```
This can be fixed by forcing the `MaxHeapSize` to something sensible. As far as I'm aware, the scripts I've included (that comes from Fabric8's Java image) set's it to 50% of the limits (with an option to specify this). So with these changes, the `MaxHeapSize` will be as below
```
    uintx MaxHeapSize                              := 132120576                           {product}
```

You can especially see increase in stability when running on lower memory allowance, for example, running Jenkins with 150mb memory limit.
before:
```
$ docker run --name jenkins  -e JAVA_OPTS='-XX:+PrintFlagsFinal -XX:+PrintGCDetails' -m=150m -d jenkins
$ docker logs jenkins 2>/dev/null | grep -i MaxHeapSize
    uintx MaxHeapSize                              := 524288000                           {product}
$ docker inspect jenkins -f '{{json .State}}'
{"Status":"exited","Running":false,"Paused":false,"Restarting":false,"OOMKilled":true,"Dead":false,"Pid":0,"ExitCode":137,"Error":"","StartedAt":"2017-03-22T23:12:52.616923385Z","FinishedAt":"2017-03-22T23:13:15.29266981Z"}
```
Notice in the output of the last command it says `"OOMKilled":true`.

Running the same with these changes does not kill it:
```
$ docker run --name jenkins-jvmfix -e JAVA_OPTS='-XX:+PrintFlagsFinal -XX:+PrintGCDetails' -m=150m -d jenkins-jvmfix
$ docker logs jenkins-jvmfix 2>/dev/null | grep -i MaxHeapSize
    uintx MaxHeapSize                              := 79691776                            {product}
$ docker inspect jenkins-jvmfix -f '{{json .State}}'
{"Status":"running","Running":true,"Paused":false,"Restarting":false,"OOMKilled":false,"Dead":false,"Pid":3505,"ExitCode":0,"Error":"","StartedAt":"2017-03-22T23:14:32.516436756Z","FinishedAt":"0001-01-01T00:00:00Z"}
```

Notice it does not kill the process due to OOM (`"OOMKilled":false`)

Of course, just like any other application that handles OOM properly, even with these changes, if the memory usage becomes too much, the kernel will still kill it. You can check this by forcing the memory limit to 50m.

The awesome guys over at Fabric8 have written the script in a way that will also handle CPU limits correctly.

Additionally, I've also moved all the shell scripts to a subdirectory, so that the repository stays cleaner.

Pull request for alpine image is: #459

### CODE REVIEW

File: scripts/java-default-options [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+# size. There is no easy solution for this, 50% seems to be are reasonable compromise.
+# However, when your container gets killed because of an OOM, then you should tune
+# the absolute values
+#
+
+# Check for memory options and calculate a sane default if not given
+max_memory() {
+  # Check whether -Xmx is already given in JAVA_OPTIONS. Then we dont
+  # do anything here
+  if echo "${JAVA_OPTIONS}" | grep -q -- "-Xmx"; then
```
Comment (hoshsadiq) [inline scripts/java-default-options]: Should be `JAVA_OPTS`

**Comment by jwhitcraft** [comment]
@hoshsadiq can this be done for the alpine image as well?

**Comment by hoshsadiq** [comment]
Didn't realise there was an alpine image. Updated!

**Comment by carlossg** [comment]
I had an implementation of this in #182 and @ndeloof had some concerns previously so I'll let him chime in
It looks good to me but it would need an entry in the README and tests, specifically checking that setting the JAVA_OPTS is honored

**Comment by ndeloof** [comment]
I indeed am concerned this get implemented in the Jenkins docker image, but this should imho be addressed in [docker-library/openjdk](https://github.com/docker-library/openjdk) as this issue for the JVM settings to match the container constraints happens in all derived images, not just here.

**Comment by carlossg** [comment]
I don't think this is ever going to be merged into the openjdk image https://github.com/docker-library/openjdk/pull/71

**Comment by carlossg** [comment]
https://github.com/docker-library/openjdk/pull/71#issuecomment-298045994

> Java 8u131 includes the -XX:+UseCGroupMemoryLimitForHeap setting (see JDK-8175898) which allows setting maximum heap size in containers without scripting, e.g.:

`java -XX:+UnlockExperimentalVMOptions -XX:+UseCGroupMemoryLimitForHeap -XX:MaxRAMFraction=1 -XshowSettings:vm -version`

**Comment by carlossg** [comment]
just need to document the jvm flags once we have the first Java 8u131 image built

**Pull request closed without merging**